### ISSUE TITLE
PutBlockListAsync throws NotSupportedException in PCL

### ISSUE THREAD

**Issue by Cheesebaron** [Azure/azure-storage-net#132]
If you use `PutBlockListAsync` in a PCL or on either Android or iOS, it will throw a `NotSupportedException` telling you:

> MD5 is not supported using the PCL

I guess this is where it goes wrong:
https://github.com/Azure/azure-storage-net/blob/preview/Lib/ClassLibraryCommon/Blob/CloudBlockBlob.cs#L2285

Here you don't check whether we are in a PCL, forcing it to calculate the MD5 and always throwing.

I see a couple of options here:

1. Provide add portable methods to calculate MD5
2. Provide an Interface that the dev implements with an MD5 implementation and supplies it at runtime, this could be done through some injection or simply passing it to the method
3. Stop forcing it to calculate MD5 in that case using a PCL
4. Let the dev do the calculation and expose the `contentMD5` as an argument in the method

Stack trace:

```
ystem.NotSupportedException: MD5 is not supported using the PCL.
05-07 10:18:29.815 I/mono-stdout( 2451): System.NotSupportedException: MD5 is not supported using the PCL.
  at Microsoft.WindowsAzure.Storage.Core.Util.MD5Wrapper..ctor () [0x00011] in <filename unknown>:0 
  at Microsoft.WindowsAzure.Storage.Core.MultiBufferMemoryStream.ComputeMD5Hash () [0x00000] in <filename unknown>:0 
  at (wrapper remoting-invoke-with-check) Microsoft.WindowsAzure.Storage.Core.MultiBufferMemoryStream:ComputeMD5Hash ()
  at Microsoft.WindowsAzure.Storage.Blob.CloudBlockBlob.PutBlockListImpl (IEnumerable`1 blocks, Microsoft.WindowsAzure.Storage.AccessCondition accessCondition, Microsoft.WindowsAzure.Storage.Blob.BlobRequestOptions options) [0x00040] in <filename unknown>:0 
  at Microsoft.WindowsAzure.Storage.Blob.CloudBlockBlob+<>c__DisplayClass30+<<PutBlockListAsync>b__2e>d__32.MoveNext () [0x00010] in <filename unknown>:0 
--- End of stack trace from previous location where exception was thrown ---
  at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw () [0x0000b] in <filename unknown>:0 
  at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess (System.Threading.Task
05-07 10:18:29.815 I/mono-stdout( 2451):   at Microsoft.WindowsAzure.Storage.Core.Util.MD5Wrapper..ctor () [0x00011] in <filename unknown>:0 
s.Task task) [0x0004e] in <filename unknown>:0 
  at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification (System.Threading.Tasks.Task task) [0x0002e] in <filename unknown>:0 
  at System.Runtime.CompilerServices.TaskAwaiter.ValidateEnd (System.Threading.Tasks.Task task) [0x0000b] in <filename unknown>:0 
  at System.Runtime.CompilerServices.ConfiguredTaskAwaitable+ConfiguredTaskAwaiter.GetResult () [0x00000] in <filename unknown>:0 
  at BK.Platform.Storage.Cloud.CloubBlockBlobExtensions+<UploadFromStreamWithContinuationAsync>d__3.MoveNext () [0x005b0] in c:\vcs\tfs\GreenSDK\Main\CloudStorage\CloudStorage\Helpers\CloubBlockBlobExtensions.cs:114 
--- End of stack trace from previous location where exception was thrown ---
  at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw () [0x0000b] in <filename unknown>:0 
  at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess (System.Threading.Tasks.Task task) [0x0004e] in <filename unknown>:0 
  at System.Runtime
05-07 10:18:29.815 I/mono-stdout( 2451):   at Microsoft.WindowsAzure.Storage.Core.MultiBufferMemoryStream.ComputeMD5Hash () [0x00000] in <filename unknown>:0 
.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification (System.Threading.Tasks.Task task) [0x0002e] in <filename unknown>:0 
  at System.Runtime.CompilerServices.TaskAwaiter.ValidateEnd (System.Threading.Tasks.Task task) [0x0000b] in <filename unknown>:0 
  at System.Runtime.CompilerServices.ConfiguredTaskAwaitable+ConfiguredTaskAwaiter.GetResult () [0x00000] in <filename unknown>:0 
  at BK.Platform.Storage.Cloud.CloudStorageProvider+<>c__DisplayClass6e+<<UploadStorageFileAsync>b__6b>d__74.MoveNext () [0x0007c] in c:\vcs\tfs\GreenSDK\Main\CloudStorage\CloudStorage\Provider\CloudStorageProvider.cs:482 
--- End of stack trace from previous location where exception was thrown ---
  at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw () [0x0000b] in <filename unknown>:0 
  at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess (System.Threading.Tasks.Task task) [0x0004e] in <filename unknown>:0 
  at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNoti
05-07 10:18:29.815 I/mono-stdout( 2451):   at (wrapper remoting-invoke-with-check) Microsoft.WindowsAzure.Storage.Core.MultiBufferMemoryStream:ComputeMD5Hash ()
[0:] System.NotSupportedException: MD5 is not supported using the PCL.
  at Microsoft.WindowsAzure.Storage.Core.Util.MD5Wrapper..ctor () [0x00011] in <filename unknown>:0 
  at Microsoft.WindowsAzure.Storage.Core.MultiBufferMemoryStream.ComputeMD5Hash () [0x00000] in <filename unknown>:0 
  at (wrapper remoting-invoke-with-check) Microsoft.WindowsAzure.Storage.Core.MultiBufferMemoryStream:ComputeMD5Hash ()
  at Microsoft.WindowsAzure.Storage.Blob.CloudBlockBlob.PutBlockListImpl (IEnumerable`1 blocks, Microsoft.WindowsAzure.Storage.AccessCondition accessCondition, Microsoft.WindowsAzure.Storage.Blob.BlobRequestOptions options) [0x00040] in <filename unknown>:0 
  at Microsoft.WindowsAzure.Storage.Blob.CloudBlockBlob+<>c__DisplayClass30+<<PutBlockListAsync>b__2e>d__32.MoveNext () [0x00010] in <filename unknown>:0 
--- End of stack trace from previous location where exception was thrown ---
  at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw () [0x0000b] in <filename unknown>:0 
  at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess (System.Threading.Tasks.Task task) [0x0004e] in <filename unknown>:0 
  at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification (System.Threading.Tasks.Task task) [0x0002e] in <filename unknown>:0 
  at System.Runtime.CompilerServices.TaskAwaiter.ValidateEnd (System.Threading.Tasks.Task task) [0x0000b] in <filename unknown>:0 
  at System.Runtime.CompilerServices.ConfiguredTaskAwaitable+ConfiguredTaskAwaiter.GetResult () [0x00000] in <filename unknown>:0 
  at BK.Platform.Storage.Cloud.CloubBlockBlobExtensions+<UploadFromStreamWithContinuationAsync>d__3.MoveNext ()
```

**Comment by pemari-msft** [comment]
Thanks for reporting this. Yes, there was an ifdef missed here. We'll fix this soon and publish it as a hotfix. Sorry for the inconvenience.

**Comment by pemari-msft** [comment]
Hi, thanks again for catching this. This should now be fixed as of version 4.4.1-preview (https://www.nuget.org/packages/WindowsAzure.Storage/).

**Comment by Cheesebaron** [comment]
Alright, going to try it out tomorrow :+1:
